### DISCUSSION TITLE
Mocking decorators on a route in a pytest

### DISCUSSION THREAD

**Discussion by andez2000** [fastapi/fastapi#9881] [Questions]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
# app code
from functools import wraps
from fastapi import HTTPException
from fastapi.routing import APIRouter


router = APIRouter()

def fake_authorize(
):
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            raise HTTPException(status_code=403, detail="Role authorization failed")
        
        return wrapper

    return decorator

@router.get(
    "/testing/hello-world-authorize",
    tags=["testing"]
)
@fake_authorize()
async def hello_world():
    return {"message": "Hello, World!"}
    

# test code
import pytest
from pytest_mock import MockerFixture
from typing import Generator
from httpx import AsyncClient
from app.main import app as main_app
from app.api.testing.simple_api_decorator import fake_authorize

@pytest.fixture
async def async_client() -> Generator:
    async with AsyncClient(app=main_app, base_url="http://testserver") as client:
        yield client


@pytest.mark.asyncio
@pytest.mark.developer
async def test_get_with_decorator(async_client: AsyncClient, mocker: MockerFixture):
    print("Running test")
    
    def mock_fake_authorize():
        print("Mock fake_authorize is called")
        return lambda func: func
    
    # \app\api\testing\simple_api_decorator.py contains the fake_authorize
    mocker.patch("app.api.testing.simple_api_decorator.fake_authorize", new=mock_fake_authorize)

    response = await async_client.get("/testing/hello-world-authorize")

    assert response.status_code == 200
    assert response.json() == {"message": "Hello, World!"}
```

### Description

I am looking at testing various scenarios in FastAPI.

In the provided code, I was hoping to patch the decorator and override the hard coded 403 to prove this can be done.  When I run the test I get the assertion error on the status code which is 403.  I want my test to assert I get a 200 response with the message.

In the real world there is various 3rd party decorators that handle this stuff for you.  But I am just trying to do the minimal code replication.

Can we override the decorators?  What is the best approach?

### Operating System

Windows

### Operating System Details

Windows 10

### FastAPI Version

0.95.0

### Python Version

3.11

**Comment by andez2000** [comment]
OK...  Having persevered a little bit more, we can make the test pass with

__api.py__

```python
from functools import wraps
from fastapi.routing import APIRouter

from sources.app.api.testing.decorators import fake_authorize


router = APIRouter()


@router.get(
    "/testing/hello-world-authorize"
)
@fake_authorize()
async def hello_world():
    return {"message": "Hello, World!"}
```

__decorators.py__

```python
def the_check():
    raise HTTPException(status_code=403, detail="Role authorization failed")

def fake_authorize(
):
    def decorator(func):
        @wraps(func)
        async def wrapper(*args, **kwargs):
            the_check()
        
        return wrapper

    return decorator
```

__test.py__

```python
import pytest
from pytest import MonkeyPatch
from typing import Generator
from httpx import AsyncClient
from sources.app.api.testing import decorators


def mock_fake_authorize():
    print("Mock fake_authorize is called")
    return lambda func: func


@pytest.fixture
async def async_client(monkeypatch: MonkeyPatch) -> Generator:
    monkeypatch.setattr(decorators, "fake_authorize", mock_fake_authorize)

    from app.main import app as main_app
    
    async with AsyncClient(app=main_app, base_url="http://testserver") as client:
        yield client


@pytest.mark.asyncio
async def test_get_with_decorator(async_client: AsyncClient):
    print("Running test")

    response = await async_client.get("/testing/hello-world-authorize")

    assert response.status_code == 200
    assert response.json() == {"message": "Hello, World!"}
```

**Comment by chandanch** [comment]
@andez2000  Thanks for this, I was stuck with it for sometime this helped, the trick here is to import app after the patch has been added else it will always call the middleware function.

Also we cannot patch within the test function as well it will still call the actual middleware

**Reply by chandanch** [reply]
If you import the app before applying the monkeypatch on the middleware, FastAPI will have already loaded the original middleware, so your patching won’t have any effect.
